### PR TITLE
Fill SROpFlash into CAFs

### PR DESCRIPTION
**Pull request by JackSmedley**
Fills new SROpFlash product into CAFs. See also pull requests in sbnanaobj (https://github.com/SBNSoftware/sbnanaobj/pull/67) and icaruscode (https://github.com/SBNSoftware/icaruscode/pull/409)

### CODE REVIEW

File: sbncode/CAFMaker/CAFMaker_module.cc [hunk 1]
```diff
@@ -1064,6 +1064,24 @@ void CAFMaker::produce(art::Event& evt) noexcept {
     }
   }
 
+  // Get all of the OpFlashes
+  std::vector<caf::SROpFlash> srflashes;
+
+  for (unsigned i_tag = 0; i_tag < pandora_tag_suffixes.size(); i_tag++) {
+    const std::string &pandora_tag_suffix = pandora_tag_suffixes[i_tag];
```
Comment (cjbackhouse) [inline sbncode/CAFMaker/CAFMaker_module.cc:1071]:
```suggestion
  for(const std::string& pandora_tag_suffix: pandora_tag_suffixes){
```

File: sbncode/CAFMaker/CAFMaker_module.cc [hunk 2]
```diff
@@ -1064,6 +1064,24 @@ void CAFMaker::produce(art::Event& evt) noexcept {
...
+    art::Handle<std::vector<recob::OpFlash>> flashes_handle;
+    GetByLabelStrict(evt, fParams.OpFlashLabel() + pandora_tag_suffix, flashes_handle);
+    // fill into event
+    if (flashes_handle.isValid()) {
+      const std::vector<recob::OpFlash> &opflashes = *flashes_handle;
+      int cryostat = ( pandora_tag_suffix.find("W") != std::string::npos );
+      for (unsigned i = 0; i < opflashes.size(); i++) {
```
Comment (cjbackhouse) [inline sbncode/CAFMaker/CAFMaker_module.cc:1078]: This can also be a range-based for loop.

File: sbncode/CAFMaker/CAFMaker_module.cc [hunk 3]
```diff
@@ -1064,6 +1064,24 @@ void CAFMaker::produce(art::Event& evt) noexcept {
...
+  std::vector<caf::SROpFlash> srflashes;
+
+  for (unsigned i_tag = 0; i_tag < pandora_tag_suffixes.size(); i_tag++) {
+    const std::string &pandora_tag_suffix = pandora_tag_suffixes[i_tag];
+    art::Handle<std::vector<recob::OpFlash>> flashes_handle;
+    GetByLabelStrict(evt, fParams.OpFlashLabel() + pandora_tag_suffix, flashes_handle);
+    // fill into event
+    if (flashes_handle.isValid()) {
+      const std::vector<recob::OpFlash> &opflashes = *flashes_handle;
+      int cryostat = ( pandora_tag_suffix.find("W") != std::string::npos );
```
Comment (cjbackhouse) [inline sbncode/CAFMaker/CAFMaker_module.cc:1077]: Gross, but presumably a pre-existing pattern. Can you at least put a `? 1 : 0;` on the end?

File: sbncode/CAFMaker/CAFMaker_module.cc [hunk 4]
```diff
@@ -1496,6 +1514,8 @@ void CAFMaker::produce(art::Event& evt) noexcept {
   rec.ncrt_hits       = srcrthits.size();
   rec.crt_tracks        = srcrttracks;
   rec.ncrt_tracks       = srcrttracks.size();
+  rec.opflashes       = srflashes;
```
Comment (cjbackhouse) [inline sbncode/CAFMaker/CAFMaker_module.cc:1517]: Again, pre-existing pattern, but why not fill straight into `rec.opflashes` above?

File: sbncode/CAFMaker/FillReco.cxx [hunk 5]
```diff
@@ -111,6 +111,46 @@ namespace caf
     srtrack.hitb.plane = track.plane2;
   }
 
+  void FillOpFlash(const recob::OpFlash &flash,
+                  int InputCryostat, 
+                  caf::SROpFlash &srflash,
+                  bool allowEmpty) {
```
Comment (cjbackhouse) [inline sbncode/CAFMaker/FillReco.cxx:117]: This parameter doesn't do anything, remove it.

File: sbncode/CAFMaker/FillReco.cxx [hunk 6]
```diff
@@ -111,6 +111,46 @@ namespace caf
...
+    srflash.PEs = std::vector<float> (flash.PEs().begin(), flash.PEs().end());
+    srflash.Cryo = InputCryostat; // 0 in SBND, 0/1 for E/W in ICARUS
+
+    // Sum over each wall, not very SBND-compliant
+    float sumEast = 0.;
+    float sumWest = 0.;
+    int countingOffset = 0;
+    if ( InputCryostat == 1 ) countingOffset += 180;
+    for ( int PMT = 0 ; PMT < 180 ; PMT++ ) {
+      if ( PMT <= 89 ) sumEast += flash.PEs().at(PMT + countingOffset);
```
Comment (cjbackhouse) [inline sbncode/CAFMaker/FillReco.cxx:132]: Ugh. Can we find these numbers out from the geometry service or something?

**Review by cjbackhouse — APPROVED** [review]
Basically fine. Minor comments, plus will need adaptation to suggested changes in the sbnanaobj PR.

File: sbncode/CAFMaker/FillReco.cxx (hunk 5 above)
Comment (JackSmedley) [reply sbncode/CAFMaker/FillReco.cxx:117]: More of me following status quo. So many functions in the file had this `allowEmpty` that I thought there must be some subtle need for it that I was missing somewhere. Happy to cut it if it's actually useless.

File: sbncode/CAFMaker/CAFMaker_module.cc (hunk 4 above)
Comment (JackSmedley) [reply sbncode/CAFMaker/CAFMaker_module.cc:1517]: Exactly as you say, pre-existing pattern. I just was not sure if I should go against the apparent grain, but I can change it.

File: sbncode/CAFMaker/CAFMaker_module.cc (hunk 3 above)
Comment (JackSmedley) [reply sbncode/CAFMaker/CAFMaker_module.cc:1077]: Yeah, I agree this is gross. Happy to add that if it makes it look prettier, but if there's a more clever way of picking out the cryostat I am all ears.

**Comment by JackSmedley** [comment]
Are we sure this should have been merged? I am still not sure about those build errors above. When I build locally, I get the same checksum error as encountered above with a message from `mrb` that says the checksum value in `sbnanaobj/StandardRecord/classes_def.xml` has been updated and says to rebuild. Rebuilding after that succeeds. I am not experienced enough with this to know if that's typical or what that means, but I just want to make sure my branch doesn't cause issues.

**Pull request merged**